src/bindings/cpp/include/key.hpp

```diff
@@ -238,6 +238,18 @@ class Key
 	ckdb::Key * key; ///< holds an elektra key
 };
 
+/**
+ * Key class returned from keysets.
+ * In particular the key name has been locked.
+ */
+class FrozenKey
```

Comment:
Which advantage does the C++ user have from this? Isn't it enough if this is a python-only thing?

Reply:
> Isn't it enough if this is a python-only thing?

That's not possible unless I custom wrap *all* keyset getters. It's slow and not maintainable. It also defeats the purpose of SWIG. If you want to make the key hashable that's the only way I'm seeing right now.

> Which advantage does the C++ user have from this?

At least they know that they can't change the key name :-)

Reply:
Maybe we should even disable the setters for the name in the FrozenKey (add them doing nothing as private).

Reply:
Sure. Go ahead.

I just had another idea. I could do it python-only if the user also wants to do some more work. Not sure it will work though. In particular the user has to convert all keys into a `FrozenKey` manually. And the FrozenKey constructor will throw if the Key isn't name locked.

So something like this:
```
keyset = kdb.KeySet(kdb.Key("user/foo"), kdb.Key("user/bar"))
key = keyset[0] # still Key
# hash(key) is None
frozenkey = FrozenKey(key) # will throw if not named lock
# hash(frozenkey) is not None

Reply:
> And the FrozenKey constructor will throw if the Key isn't name locked.

Why not lock the key there?

@PhilippGackstatter can you give some input here? (as you recently made something similar in Rust)

Reply:
I'll leave this for today. In general I think the current approach is better in terms of usability as in "It just works". However since the *only* usecase is to make the key hashable, manual conversation might be the better approach. *Unless* the FrozenKey approach won't be extended. Because it gets messy really fast. Like `kdb.KeySet(1, FrozenKey(kdb.Key("user/foo")))[0]` is a kdb.Key again.

Reply:
Wouldn't it be possible in python to dynamically use the correct class depending on if the lock bit is set?

Reply:
Yes, but it means you need to wrap *every* c++ function which returns a key. And that's  slow + huge mainainence effort. Don't like the Idea at all.

Reply:
What about making FrozenKey only a typedef in C++? (Could be in an `#ifdef` that SWIG sees something else.)

Reply:
Thought about this myself. Doable but also horrible. 

I just found: https://stackoverflow.com/questions/28517290/python-classes-with-hash-support-depending-on-the-instance

What about doing the same: Throwing in `__hash__` and move on. L

Reply:
Thought about this myself. Doable but also horrible. 

I just found: https://stackoverflow.com/questions/28517290/python-classes-with-hash-support-depending-on-the-instance

What about doing the same: Throwing in `__hash__` and move on. L

Reply:
Yes, sounds like a simple and straightforward solution. (If not locked, throw in `__hash__`). Together with some way to lock the keys (at least at construction) should be everything we want about `__hash__`.

The important part here is anyway that this is properly described how to use it in the tutorial.

Reply:
> @PhilippGackstatter can you give some input here? (as you recently made something similar in Rust)

In the Rust bindings, calling `set_name` on a key in a KeySet returns an error. 

There's also a `ReadOnly` wrapper that returns a key that only implements `get_`-methods. For example, the Rust equivalent of `keyGetMeta` returns such a `ReadOnly` key. With that it is a compile time error to call `set_name` on that key.

However, this is probably not the right approach for Python.